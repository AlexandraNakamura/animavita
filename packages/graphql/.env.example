# Environment
NODE_ENV=

# Commit sha
COMMIT_SHA=

# Local server only
GRAPHQL_PORT=

# Front Security
JWT_KEY=

# AWS
AWS_S3_BUCKET_NAME=
AWS_REGION=
AWS_CLOUD_FORMATION_STACK_NAME=
