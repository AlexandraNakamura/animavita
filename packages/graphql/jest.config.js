const pkg = require('./package');

module.exports = {
  rootDir: './',
  name: pkg.name,
  displayName: pkg.name.toUpperCase(),
  testPathIgnorePatterns: ['/node_modules/', './build'],
  coverageReporters: ['lcov', 'html'],
  globalSetup: '<rootDir>/tests/setup.js',
  globalTeardown: '<rootDir>/tests/teardown.js',
  testEnvironment: '<rootDir>/tests/environment/mongodb',
  resetModules: false,
  reporters: ['default'],
  transform: {
    '^.+\\.(js|jsx|ts|tsx)?$': '<rootDir>/tests/babel.transformer',
  },
  testRegex: '(/__tests__/.*|(\\.|/)(test|spec))\\.(js|jsx|ts|tsx)?$',
  moduleFileExtensions: ['ts', 'js', 'tsx'],
};
