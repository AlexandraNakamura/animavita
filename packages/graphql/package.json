{
  "name": "@animavita/graphql",
  "version": "1.0.0",
  "private": true,
  "scripts": {
    "build": "webpack --config webpack.prod.js --display-error-details --display-modules",
    "start": "webpack --config webpack.dev.js --progress",
    "test": "jest",
    "test:watch": "jest --watch"
  },
  "dependencies": {
    "@koa/router": "^8.0.6",
    "aws-serverless-koa": "^1.0.1",
    "core-js": "^3.6.4",
    "dataloader": "^2.0.0",
    "dotenv-safe": "^8.2.0",
    "env-var": "^5.2.0",
    "graphql": "^14.5.8",
    "graphql-playground-middleware": "^1.1.2",
    "isomorphic-fetch": "^2.2.1",
    "koa": "^2.11.0",
    "koa-bodyparser": "^4.2.1",
    "koa-convert": "^1.2.0",
    "koa-cors": "^0.0.16",
    "koa-graphql": "^0.8.0",
    "koa-logger": "^3.2.1",
    "mongoose": "^5.8.9"
  },
  "devDependencies": {
    "@types/dotenv-safe": "^8.1.0",
    "@types/koa": "^2.11.0",
    "@types/koa-bodyparser": "^4.3.0",
    "@types/koa-convert": "^1.2.3",
    "@types/koa-cors": "^0.0.0",
    "@types/koa-graphql": "^0.8.3",
    "@types/koa-logger": "^3.1.1",
    "@types/koa-multer": "^1.0.0",
    "@types/koa__router": "^8.0.2",
    "@types/mongoose": "^5.5.43",
    "@types/node": "^13.5.0",
    "babel-jest": "^25.1.0",
    "babel-loader": "^8.0.6",
    "jest": "24.9.0",
    "reload-server-webpack-plugin": "^1.0.1",
    "terser-webpack-plugin": "^2.3.4",
    "webpack": "^4.41.5",
    "webpack-cli": "^3.3.10",
    "webpack-merge": "^4.2.2",
    "webpack-node-externals": "^1.7.2"
  }
}
