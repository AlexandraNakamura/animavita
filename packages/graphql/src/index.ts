import serverless from 'aws-serverless-koa';

import app from './app';
import connectDatabase from './common/database';

(async () => {
  await connectDatabase();
})();

export const handler = serverless(app);
