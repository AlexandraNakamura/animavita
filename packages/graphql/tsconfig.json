{
  "compilerOptions": {
    "rootDir": ".",
    "composite": true,
    "noEmit": true,
    "typeRoots": ["@types"]
  },
  "extends": "../../tsconfig.json",
  "include": [
    "./",
    "../../node_modules/@types",
    "./@types/mongodb.d.ts",
    "../../node_modules/jest-fetch-mock/types/index.d.ts"
  ]
}
