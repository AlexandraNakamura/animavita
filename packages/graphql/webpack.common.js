const path = require('path');

const webpack = require('webpack');
const TerserPlugin = require('terser-webpack-plugin');
const WebpackNodeExternals = require('webpack-node-externals');

module.exports = {
  target: 'node',
  optimization: {
    minimizer: [new TerserPlugin()],
  },
  externals: [
    WebpackNodeExternals({
      whitelist: ['webpack/hot/poll?1000'],
    }),
    WebpackNodeExternals({
      modulesDir: path.resolve(__dirname, '../../node_modules'),
      whitelist: [/@animavita/],
    }),
    {'aws-sdk': 'aws-sdk'},
  ],
  resolve: {
    extensions: ['.ts', '.tsx', '.js', '.json', '.mjs'],
  },
  output: {
    path: path.resolve(__dirname, 'build'),
    filename: 'index.js',
    library: 'index',
    libraryTarget: 'commonjs2',
  },
  plugins: [
    new webpack.DefinePlugin({
      'process.env.BROWSER': false,
      __DEV__: process.env.NODE_ENV !== 'production',
    }),
  ],
  module: {
    rules: [
      {
        test: /\.mjs$/,
        include: /node_modules/,
        type: 'javascript/auto',
      },
      {
        test: /\.([jt]sx?)$/,
        loader: 'babel-loader',
        options: {
          configFile: './babel.config.js',
        },
        exclude: /node_modules/,
        include: [path.resolve(__dirname, 'src'), path.resolve(__dirname, '../')],
      },
    ],
  },
};
