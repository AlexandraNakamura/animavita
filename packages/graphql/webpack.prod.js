const merge = require('webpack-merge');

const common = require('./webpack.common.js');

module.exports = merge(common, {
  mode: 'production',
  entry: {
    index: './src/index.ts',
    worker: './src/worker.ts',
  },
});
