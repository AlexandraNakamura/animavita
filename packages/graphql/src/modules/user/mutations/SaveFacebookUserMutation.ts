import {mutationWithClientMutationId} from 'graphql-relay';
import {GraphQLInt, GraphQLList, GraphQLNonNull, GraphQLString} from 'graphql';
import imageToBase64 from 'image-to-base64';
import AWS from 'aws-sdk';

import {AWS_ACCESS_KEY_ID, AWS_REGION, AWS_S3_BUCKET_NAME, AWS_SECRET_ACCESS_KEY} from '../../../common/config';
import {GraphQLContext} from '../../../types';
import UserModel, {IProfileImage, IUser, IUserDocument} from '../UserModel';
import UserType from '../UserType';

interface SaveFacebookUserMutationArgs {
  token: string;
  expires: number;
  permissions: string[];
}
interface UserIncomplete {
  id: string;
  name: string;
  email: string;
}

export default mutationWithClientMutationId({
  name: 'SaveFacebookUser',
  description: 'Save facebook user data',
  inputFields: {
    token: {type: GraphQLNonNull(GraphQLString)},
    expires: {type: GraphQLNonNull(GraphQLInt)},
    permissions: {type: GraphQLNonNull(GraphQLList(GraphQLNonNull(GraphQLString)))},
  },
  mutateAndGetPayload: async (args: SaveFacebookUserMutationArgs, ctx: GraphQLContext) => {
    const {token} = args;

    const userIncomplete: UserIncomplete = await (
      await fetch(`https://graph.facebook.com/me?access_token=${token}&fields=id,name,email`)
    ).json();

    if (!userIncomplete || !userIncomplete.email) return {error: 'Failed to fetch basic user data'};

    const {url: profileUrl} = await fetch(
      `https://graph.facebook.com/${userIncomplete.id}/picture?height=720&width=720`,
    );

    const {id, name, email} = userIncomplete;

    let newUserDocument: IUser = {
      ids: [{id, providedBy: 'facebook'}],
      name,
      emails: [{email, providedBy: 'facebook'}],
    };

    const checkIfUserAlreadyExistsPipeline = [
      {
        $match: {
          $or: [{ids: {$in: newUserDocument.ids}}, {emails: {$in: newUserDocument.emails}}],
        },
      },
    ];

    const userAggregation = await UserModel.aggregate(checkIfUserAlreadyExistsPipeline);

    // if the user provider already exists
    if (userAggregation.length && userAggregation.length === 1) {
      let dbUser: IUserDocument = userAggregation[0];

      // verify if name was updated
      if (name.length > dbUser.name.length) {
        await UserModel.updateOne({_id: dbUser._id}, {$set: {name}});
        dbUser = await UserModel.findById(dbUser._id).lean()!;
      }

      dbUser = await saveOrUpdateProfileImage(dbUser, profileUrl, userIncomplete);

      return {
        user: {
          ...dbUser,
          providerIds: dbUser.ids,
        },
      };
    } else {
      if (profileUrl) {
        const {Location, Key} = await uploadProfileImage(profileUrl, userIncomplete);

        newUserDocument.profileImages = [
          {
            key: Key,
            location: Location,
            originUri: profileUrl,
            providedBy: 'facebook',
          },
        ];
      }

      // @ts-ignore
      const user = (await new UserModel(newUserDocument).save())._doc;

      return {
        user: {
          ...user,
          providerIds: user.ids,
        },
      };
    }
  },
  outputFields: {
    error: {
      type: GraphQLString,
      resolve: obj => obj.error,
    },
    user: {
      type: UserType,
      resolve: obj => obj.user,
    },
  },
});

async function uploadProfileImage(url: string, user: UserIncomplete) {
  const base64Image = await imageToBase64(url);

  AWS.config.update({accessKeyId: AWS_ACCESS_KEY_ID, secretAccessKey: AWS_SECRET_ACCESS_KEY, region: AWS_REGION});

  const buffer = Buffer.from(base64Image, 'base64');

  const s3 = new AWS.S3();

  const params = {
    Bucket: AWS_S3_BUCKET_NAME,
    Key: `${user.id}${new Date().getTime()}.jpeg`,
    Body: buffer,
    ACL: 'public-read',
    ContentEncoding: 'base64',
    ContentType: 'image/jpeg',
  };

  return await s3.upload(params).promise();
}

async function saveOrUpdateProfileImage(dbUser: IUserDocument, profileUrl: string, userIncomplete: UserIncomplete) {
  let shouldUpdateProfileImage = false;
  // if user doesnâ€™t have profile image yet, upload it
  if (!dbUser.profileImages || dbUser.profileImages.length === 0) {
    shouldUpdateProfileImage = true;
  } else {
    // filter by facebook images
    const fbImages = dbUser.profileImages.filter(profileImage => profileImage.providedBy === 'facebook');

    // user might have profile image from another provider, use the image from last provider used
    if (!fbImages || fbImages.length === 0) {
      shouldUpdateProfileImage = true;
    } else {
      let alreadyExitsOnS3 = false;

      for await (const fbImage of fbImages) {
        const originBase64Image = await imageToBase64(fbImage.originUri);
        const fbBase64Image = await imageToBase64(fbImage.location);

        // verify if the current profile image already exists on S3
        if (fbBase64Image == originBase64Image) {
          alreadyExitsOnS3 = true;
        }
      }

      if (!alreadyExitsOnS3) {
        shouldUpdateProfileImage = true;
      }
    }
  }

  if (shouldUpdateProfileImage) {
    const {Location, Key} = await uploadProfileImage(profileUrl, userIncomplete);

    const fbProfileImage: IProfileImage = {
      key: Key,
      location: Location,
      originUri: profileUrl,
      providedBy: 'facebook',
    };

    let profileImages: IProfileImage[] = [];
    if (dbUser.profileImages) {
      profileImages.push(...dbUser.profileImages, fbProfileImage);
    } else {
      profileImages.push(fbProfileImage);
    }

    await UserModel.updateOne(
      {_id: dbUser._id},
      {
        $set: {profileImages},
      },
    );
    dbUser = await UserModel.findById(dbUser._id).lean()!;
  }
  return dbUser;
}
