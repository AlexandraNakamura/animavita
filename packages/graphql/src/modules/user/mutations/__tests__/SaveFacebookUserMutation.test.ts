import {graphql} from 'graphql';

import {connectMongoose, disconnectMongoose, restartCounters, gql, getContext} from '../../../../../tests/helpers';
import schema from '../../../../schema';

import {SaveFacebookUserMutationArgs} from '../SaveFacebookUserMutation';

beforeAll(connectMongoose);
beforeEach(restartCounters);
afterAll(disconnectMongoose);

interface MutationVariables {
  input: SaveFacebookUserMutationArgs;
}

const mutation = gql`
  mutation M($input: SaveFacebookUserInput!) {
    SaveFacebookUser(input: $input) {
      clientMutationId
      error
      token
      user {
        name
        emails {
          email
          providedBy
        }
        providerIds {
          id
          providedBy
        }
      }
    }
  }
`;

describe('SaveFacebookUserMutation', () => {
  it('should create a new account if never logged in with facebook before', async () => {
    // mock fetch of UserIncomplete
    // @ts-ignore
    fetch.once(JSON.stringify({id: 'fakeid', name: 'Fake Name', email: 'fakeemail@fake.com'}));

    // mock fetch of user profile
    // @ts-ignore
    fetch.once(JSON.stringify({url: 'https://fakeprofileurl.com'}));

    const rootValue = {};
    const context = await getContext();

    const variables: MutationVariables = {
      input: {
        token: 'somefaketoken',
        expires: 5797,
        permissions: ['email', 'profile'],
      },
    };

    const result = await graphql(schema, mutation, rootValue, context, variables);

    expect(result.errors).toBeUndefined();
    expect(result.data!.SaveFacebookUser.error).toBeNull();
    expect(result.data!.SaveFacebookUser.token).toBeDefined();
    expect(result.data!.SaveFacebookUser.user).toMatchSnapshot();
  });
});
