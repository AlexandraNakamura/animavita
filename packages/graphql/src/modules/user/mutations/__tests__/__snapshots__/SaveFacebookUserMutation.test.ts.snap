// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SaveFacebookUserMutation should add facebook info if user already logged in with another provider (and update name if its longer) 1`] = `
Object {
  "emails": Array [
    Object {
      "email": "fakeemail@fake.com",
      "providedBy": "google",
    },
  ],
  "name": "Fake Loooooooonger Name",
  "profileImages": Array [
    Object {
      "providedBy": "facebook",
      "url": "https://fakebucket.com",
    },
  ],
  "providerIds": Array [
    Object {
      "id": "googlefakeid",
      "providedBy": "google",
    },
    Object {
      "id": "facebookfakeid",
      "providedBy": "facebook",
    },
  ],
}
`;

exports[`SaveFacebookUserMutation should create a new account if never logged in with facebook before 1`] = `
Object {
  "emails": Array [
    Object {
      "email": "fakeemail@fake.com",
      "providedBy": "facebook",
    },
  ],
  "name": "Fake Name",
  "profileImages": Array [
    Object {
      "providedBy": "facebook",
      "url": "https://fakebucket.com",
    },
  ],
  "providerIds": Array [
    Object {
      "id": "fakeid",
      "providedBy": "facebook",
    },
  ],
}
`;
